FreeRTOS/Test/Target/tests/smp/fr09_interruptWaitCritical/test.c

```diff
@@ -0,0 +1,216 @@
+    {
+        uxNumTasksRunning = uxTaskGetSystemState( ( TaskStatus_t * const ) &taskStatus, uxTaskStatusArraySize, &ulTotalRunTime );
+
+        for( uxIdx = 0; uxIdx < uxNumTasksRunning; uxIdx++ )
+        {
+            if( ( strcmp( taskStatus[ uxIdx ].pcTaskName, "TaskA" ) == 0 ) && ( taskStatus[ uxIdx ].eCurrentState == eRunning ) )
+            {
+                xTaskARan = pdTRUE;
+            }
+
+            if( ( strcmp( taskStatus[ uxIdx ].pcTaskName, "TaskB" ) == 0 ) && ( taskStatus[ uxIdx ].eCurrentState == eRunning ) )
+            {
+                xTaskBRan = pdTRUE;
+            }
+
+            if( ( strcmp( taskStatus[ uxIdx ].pcTaskName, "TaskC" ) == 0 ) && ( taskStatus[ uxIdx ].eCurrentState == eRunning ) )
+            {
+                xTaskCRan = pdTRUE;
+            }
+        }
+
+        if( ( xTaskARan == pdTRUE ) && ( xTaskCRan == pdTRUE ) && ( xTaskBRan == pdTRUE ) )
+        {
+            xAllTasksHaveRun = pdTRUE;
+        }
+
+        ulTaskSwitchCount++;
+
+        if( ulTaskSwitchCount > 1500 )
+        {
+            xTestFailed = pdTRUE;
+        }
+    }
+}
+/*-----------------------------------------------------------*/
+
+static void fr09_validateAllTasksHaveRun( void )
+{
+    xTaskCreateAffinitySet( vPrvTaskA, "TaskA", configMINIMAL_STACK_SIZE, NULL,
+                            mainTASK_A_PRIORITY, 0x2, &xTaskAHandler );
+
+    xTaskCreate( vPrvTaskB, "TaskB", configMINIMAL_STACK_SIZE, NULL,
+                 mainTASK_B_PRIORITY, &xTaskBHandler );
+
+    xTaskCreate( vPrvTaskC, "TaskC", configMINIMAL_STACK_SIZE, NULL,
+                 mainTASK_C_PRIORITY, &xTaskCHandler );
+
+    while( !xAllTasksHaveRun )
+    {
+        vTaskDelay( pdMS_TO_TICKS( 100 ) );
+    }
+
+    TEST_ASSERT_TRUE( xTaskBHasEnteredCriticalSection == pdFALSE );
+}
+/*-----------------------------------------------------------*/
+
+static void vPrvTaskA( void * pvParameters )
+{
+    taskENTER_CRITICAL();
+    xTaskAHasEnteredCriticalSection = pdTRUE;
+    vPortBusyWaitMicroseconds( ( uint32_t ) 250000 );
+    // xTaskNotify( xTaskBHandler, 0, eNoAction );
+    vTaskSuspend( xTaskBHandler );
+    vPortBusyWaitMicroseconds( ( uint32_t ) 10000000 );
+    taskEXIT_CRITICAL();
+    xTaskAHasExitedCriticalSection = pdTRUE;
+
+    /* idle the task */
+    for( ; ; )
+    {
+        vTaskDelay( pdMS_TO_TICKS( 10 ) );
+    }
+}
+/*-----------------------------------------------------------*/
+
+static void vPrvTaskB( void * pvParameters )
+{
+    vTaskDelay( pdMS_TO_TICKS( 10 ) );
+
+    taskENTER_CRITICAL();
+    xTaskBHasEnteredCriticalSection = pdTRUE;
+    vPortBusyWaitMicroseconds( ( uint32_t ) 8000000 );
+    taskEXIT_CRITICAL();
+
+    /* idle the task */
+    for( ; ; )
+    {
+        vTaskDelay( pdMS_TO_TICKS( 10 ) );
+    }
+}
+/*-----------------------------------------------------------*/
+
+static void vPrvTaskC( void * pvParameters )
+{
+    vPortBusyWaitMicroseconds( ( uint32_t ) 250000 );
```

Comment:
We should keep task C in busy loop all the time to make sure it doesn't release the MCU resource to other tasks.